lib/osbuild.py

```diff
@@ -0,0 +1,370 @@
+            datastream = "{ds_file}"
+        ''') + '\n'
+
+    def add_openscap_tailoring(self, *, selected=None, unselected=None):
+        if '[customizations.openscap.tailoring]' in self.assembled:
+            raise SyntaxError("openscap.taioring section already exists")
+        if not selected and not unselected:
+            return
+        self.assembled += '[customizations.openscap.tailoring]\n'
+        for name, vals in [('selected', selected), ('unselected', unselected)]:
+            if vals:
+                strings = ','.join(f'"{x}"' for x in vals)
+                self.assembled += f'{name} = [ {strings} ]\n'
+
+    @contextlib.contextmanager
+    def to_tmpfile(self):
+        bp = self.assembled
+        util.log(f"using blueprint:\n{textwrap.indent(bp, '    ')}")
+        with tempfile.NamedTemporaryFile(mode='w') as f:
+            f.write(bp)
+            f.flush()
+            yield Path(f.name)
+
+    @contextlib.contextmanager
+    def to_composer(self):
+        blueprints = composer_cli_out('blueprints', 'list', log=False)
+        if self.NAME in blueprints:
+            composer_cli('blueprints', 'delete', self.NAME)
+        with self.to_tmpfile() as f:
+            composer_cli('blueprints', 'push', f)
+        yield self.NAME
+        composer_cli('blueprints', 'delete', self.NAME)
+
+
+class Guest(virt.Guest):
+    """
+    TODO document
+    """
+    DATASTREAM = Path(util.RPMPACK_DATA) / 'ds.xml'
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.osbuild_log = f'{virt.GUEST_IMG_DIR}/{self.name}-osbuild.txt'
+
+    def wipe(self):
+        super().wipe()
+        log = Path(self.osbuild_log)
+        if log.exists():
+            os.remove(log)
+
+    def install(*args, **kwargs):
+        raise NotImplementedError("install() is not supported, use create()")
+
+    def create(self, *, blueprint=None, bp_verbatim=None, profile=None):
+        """
+        ... creates an image .. imports via virt-install
+        """
+        util.log(f"creating guest {self.name}")
+
+        # remove any previously installed guest
+        self.wipe()
+
+        if not blueprint:
+            blueprint = Blueprint()
+
+        if not bp_verbatim:
+            blueprint.add_host_repositories()
+            # copy our default package list from virt.Kickstart
+            for pkg in virt.Kickstart.PACKAGES:
+                blueprint.add_package(pkg)
+            # copy partitions too
+            for mountpoint, size in virt.Kickstart.PARTITIONS:
+                blueprint.add_partition(mountpoint, size*1024*1024)
+            # generate the ssh key on the same place as virt.Guest
+            util.ssh_keygen(self.ssh_keyfile_path)
+            with open(f'{self.ssh_keyfile_path}.pub') as f:
+                pubkey = f.read().rstrip()
+            blueprint.add_user('root', password=virt.GUEST_LOGIN_PASS, ssh_pubkey=pubkey)
```

Comment:
Does it work? I expect `root` is there by default and thus customization won't add/change it.

That's what came to mind when seeing [4.8.6. Setting SSH key for existing users](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/composing_a_customized_rhel_system_image/creating-system-images-with-composer-command-line-interface_composing-a-customized-rhel-system-image#setting-ssh-key-for-existing-users_creating-system-images-with-composer-command-line-interface)

Reply:
And the password won't work as now you add plaintext but hash is expected.

Reply:
It does indeed work. I used to have a password crypt function (as evidenced by a leftover comment I'll remove during cleanup/documentation in later commits), but then I saw Blueprint examples with plaintexts and I switched back.

Then I read the docs on https://www.osbuild.org/guides/image-builder-on-premises/blueprint-reference.html#additional-users and it said
```
Note: If the password starts with $6$, $5$, or $2b$ it will be stored
as an encrypted password. Otherwise it will be treated as a plain text
password.
```

As for the ssh key, I also asked osbuild developers, and was told that it does the same thing as `[[customizations.sshkey]]` which is an older customization method, and that both are internally merged during blueprint parsing, and was recommended to use `[[customizations.user]]`.